File: apps/web/tests/fixtures/visualizations.ts
import type { HomeVisualizationsResponse, MetricSeriesVisualization } from "../../types/visualization";

const period = { start: "2026-08-02T07:00:00Z", end: "2026-09-01T07:00:00Z" };
const source = ["synthetic:test-sp-api"];
const evidence = [{ kind: "TOOL_OUTPUT" as const, reference_id: "tool:get_metric_series:test" }];

function metricSeries(metric: string, unit: string, base: number, step: number): MetricSeriesVisualization {
  return {
    metric,
    scope: "STORE",
    unit,
    lookback_days: 30,
    maturity: metric.startsWith("ad_") || ["acos", "cpc", "ctr"].includes(metric) ? "PROVISIONAL" : "MATURED",
    points: Array.from({ length: 30 }, (_, index) => ({
      period: `2026-08-${String(index + 2).padStart(2, "0")}`,
      value: Number((base + step * index + ((index % 5) - 2) * step * 0.35).toFixed(2)),
    })),
    evidence_refs: evidence,
    data_period: period,
    source,
    updated_at: "2026-08-31T12:00:00Z",
    confidence: 0.96,
    limitations: [],
    synthetic: true,
  };
}

export function homeVisualizations(): HomeVisualizationsResponse {
  return {
    business_date: "2026-08-31",
    marketplace: "ATVPDKIKX0DER",
    lookback_days: 30,
    metric_series: [
      metricSeries("orders", "COUNT", 35, 0.4),
      metricSeries("sales", "USD", 1320, 13),
      metricSeries("sessions", "COUNT", 510, 3.2),
      metricSeries("cvr", "PERCENT", 7.1, 0.035),
      metricSeries("ad_spend", "USD", 61, 0.32),
      metricSeries("ad_sales", "USD", 106, 1.1),
      metricSeries("acos", "PERCENT", 57, -0.18),
      metricSeries("cpc", "USD_PER_CLICK", 1.24, -0.004),
      metricSeries("ctr", "PERCENT", 0.43, 0.003),
    ],
    top_entities: [{
      metric: "sales",
      scope: "STORE",
      entity_type: "ASIN",
      unit: "USD",
      lookback_days: 30,
      entities: [
        { rank: 1, entity_id: "B0TEST001", label: "Cooling Memory Foam Pillow", value: 8420 },
        { rank: 2, entity_id: "B0TEST002", label: "Adjustable Sleep Pillow", value: 6250 },
        { rank: 3, entity_id: "B0TEST003", label: "Bamboo Pillowcase Set", value: 3910 },
      ],
      evidence_refs: evidence,
      data_period: period,
      source,
      updated_at: "2026-08-31T12:00:00Z",
      confidence: 0.97,
      limitations: [],
      synthetic: true,
    }],
    mix_breakdowns: [{
      metric: "sales",
      scope: "STORE",
      entity_type: "ASIN",
      unit: "USD",
      lookback_days: 30,
      total: 20780,
      categories: [
        { label: "Cooling Memory Foam Pillow", value: 8420, share_pct: 40.52, entity_id: "B0TEST001" },
        { label: "Adjustable Sleep Pillow", value: 6250, share_pct: 30.08, entity_id: "B0TEST002" },
        { label: "Bamboo Pillowcase Set", value: 3910, share_pct: 18.82, entity_id: "B0TEST003" },
        { label: "Other", value: 2200, share_pct: 10.58 },
      ],
      evidence_refs: evidence,
      data_period: period,
      source,
      updated_at: "2026-08-31T12:00:00Z",
      confidence: 0.97,
      limitations: [],
      synthetic: true,
    }],
    synthetic: true,
  };
}

File: apps/api/src/amazon_ai_api/models/visualizations.py
from __future__ import annotations

from datetime import date, datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field

from amazon_ai_api.models.home import EvidenceReference
from amazon_ai_api.models.provenance import DataPeriod
from amazon_ai_api.models.tools import StoreToolInput


class VisualizationScope(StrEnum):
    STORE = "STORE"


class MetricName(StrEnum):
    ORDERS = "orders"
    SALES = "sales"
    SESSIONS = "sessions"
    CVR = "cvr"
    AD_SPEND = "ad_spend"
    AD_SALES = "ad_sales"
    ACOS = "acos"
    CPC = "cpc"
    CTR = "ctr"
    INVENTORY_DAYS = "inventory_days"
    CONTRIBUTION_PROFIT = "contribution_profit"


class EntityMetric(StrEnum):
    ORDERS = "orders"
    SALES = "sales"
    SESSIONS = "sessions"


class MixMetric(StrEnum):
    ORDERS = "orders"
    SALES = "sales"


class EntityType(StrEnum):
    ASIN = "ASIN"


class MetricSeriesToolInput(StoreToolInput):
    metric: MetricName
    scope: VisualizationScope = VisualizationScope.STORE
    lookback_days: int = Field(default=30, ge=1, le=90)


class TopEntitiesToolInput(StoreToolInput):
    metric: EntityMetric
    scope: VisualizationScope = VisualizationScope.STORE
    entity_type: EntityType = EntityType.ASIN
    lookback_days: int = Field(default=30, ge=1, le=90)
    limit: int = Field(default=5, ge=1, le=5)


class MixBreakdownToolInput(StoreToolInput):
    metric: MixMetric
    scope: VisualizationScope = VisualizationScope.STORE
    entity_type: EntityType = EntityType.ASIN
    lookback_days: int = Field(default=30, ge=1, le=90)
    max_slices: int = Field(default=5, ge=2, le=5)


class TimeSeriesPoint(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    period: date
    value: float


class MetricSeriesPayload(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    metric: MetricName
    scope: VisualizationScope
    unit: str
    lookback_days: int = Field(ge=1, le=90)
    maturity: str
    points: tuple[TimeSeriesPoint, ...]


class RankedEntity(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    rank: int = Field(ge=1, le=5)
    entity_id: str = Field(min_length=1)
    label: str = Field(min_length=1)
    value: float = Field(ge=0)


class TopEntitiesPayload(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    metric: EntityMetric
    scope: VisualizationScope
    entity_type: EntityType
    unit: str
    lookback_days: int = Field(ge=1, le=90)
    entities: tuple[RankedEntity, ...]


class MixCategory(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    label: str = Field(min_length=1)
    value: float = Field(ge=0)
    share_pct: float = Field(ge=0, le=100)
    entity_id: str | None = None


class MixBreakdownPayload(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    metric: MixMetric
    scope: VisualizationScope
    entity_type: EntityType
    unit: str
    lookback_days: int = Field(ge=1, le=90)
    total: float = Field(gt=0)
    categories: tuple[MixCategory, ...] = Field(min_length=1, max_length=5)


class EvidenceBackedVisualization(BaseModel):
    model_config = ConfigDict(extra="forbid")

    evidence_refs: tuple[EvidenceReference, ...] = Field(min_length=1)
    data_period: DataPeriod
    source: tuple[str, ...] = Field(min_length=1)
    updated_at: datetime
    confidence: float = Field(ge=0, le=1)
    limitations: tuple[str, ...]
    synthetic: bool


class MetricSeriesVisualization(EvidenceBackedVisualization):
    metric: MetricName
    scope: VisualizationScope
    unit: str
    lookback_days: int = Field(ge=1, le=90)
    maturity: str
    points: tuple[TimeSeriesPoint, ...] = Field(min_length=1)


class TopEntitiesVisualization(EvidenceBackedVisualization):
    metric: EntityMetric
    scope: VisualizationScope
    entity_type: EntityType
    unit: str
    lookback_days: int = Field(ge=1, le=90)
    entities: tuple[RankedEntity, ...] = Field(min_length=1, max_length=5)


class MixBreakdownVisualization(EvidenceBackedVisualization):
    metric: MixMetric
    scope: VisualizationScope
    entity_type: EntityType
    unit: str
    lookback_days: int = Field(ge=1, le=90)
    total: float = Field(gt=0)
    categories: tuple[MixCategory, ...] = Field(min_length=1, max_length=5)


class HomeVisualizations(BaseModel):
    model_config = ConfigDict(extra="forbid")

    business_date: date
    marketplace: str
    lookback_days: int = Field(ge=1, le=90)
    metric_series: tuple[MetricSeriesVisualization, ...]
    top_entities: tuple[TopEntitiesVisualization, ...]
    mix_breakdowns: tuple[MixBreakdownVisualization, ...]
    synthetic: bool
